src/std/net/socket/server-test.ss

```diff
@@ -0,0 +1,75 @@
+;;; -*- Gerbil -*-
+;;; (C) vyzo at hackzen.org
+;;; Socket server tests
+
+(import :gerbil/gambit/threads
+        :std/misc/threads
+        :std/net/socket
+        :std/net/socket/base
+        :std/os/socket
+        :std/error
+        :std/sugar
+        :std/test)
+(export socket-server-test)
+
+(def socket-server-test
+  (test-suite "test :std/net/socket"
+
+    (def server-address "localhost:4242")
+
+    (def (echo sock)
+      (def buf (make-u8vector 256))
+      (try
+       (let lp ()
+         (def len (ssocket-recv sock buf))
```

Comment:
you should check this for len being 0, denoting a remotely closed socket.

Reply:
Added the check, and called `ssocket-close` on the server side in this case. I wasn't sure if this was necessary but I noticed that the entire test would hang. Interrupting it showed that the thread was blocked in `kevent`. Perhaps I'm doing something subtly wrong in the shutdown phase?

Reply:
well, we do want to close the socket when the other side has shutdown,  so that it's closed promptly (otherwise it will take until the garbage collector sweep).
Not sure why it hangs if you don't close, probably because of some subtlety in `kqeueue` behaviour? The socket is still in the event set after all.

Reply:
So just to make sure, is the `kqueue` fd set to nonblocking? 
It really shouldn't hang without closing the socket, as the garbage collector should close it on its own when it becomes unreachable.

Reply:
Maybe so, although I would double-check the nonblocking nature of the `kqueue` fd and make sure there is a will for the socket in `kqueue-server` so that it gets properly removed from the event set when it is deemed unreachable by the gc.

Reply:
also, maybe the gc never gets to run -- with the socket lingering on in the event set.

Reply:
Yes, and I think I might need to be sure to close the main server socket (created by `ssocket-listen`), as well.

Reply:
So looking at the kqueue code, the file descriptor is not set to nonblocking.
It's possible that we have spurious interrupt in the wait, and then the subsequent `kqueue` call blocks.

Can you add an `fd-set-nonblock` (and `fd-set-closeonexec`) from `:std/os/fcntl` in the raw device created by `kqueue`? I think this might solve the problem, as we really shouldn't have to close the sockets for kqueue to not block.

Note that epoll has the same mistake, it should have a set-nonblock and set-closeonexec as well.

Reply:
Fixing this, will open pr (and you can rebase).

Reply:
Ok, fixed in #100.

Reply:
Ah, I think i found the issue -- and it's probably not the nonblocking stuff.
You are doing a kqueue-wait with no timeout, which will block the process indefinitely.
This should really be done with a timeout of 0, which coupled with the poll on the kqueue file descriptor, should provide asynchronous semantics without blocking the process.

Reply:
So the nonblock is useless, the key is the zero timeout for `kevent` :)

Reply:
reverted the useless nonblock in #101...

Reply:
What we need to do is fix `kqueue-wait` to have a zero timeout -- this should definitively fix the "blocked on kqueue" issue.

Reply:
better to rename to `kqueue-poll` i think.

Reply:
Thanks for digging into this one. Will push this ASAP.

Reply:
let's rebase for #102, so that we can verify and merge!